tools/schemacode/bidsschematools/validator.py

```diff
@@ -38,285 +146,224 @@ def _get_paths(
-    '(?P<run>[0-9]+)'
-    >>> _capture_regex("run", "[0-9]+", True)
-    '(?P=run)'
-    >>> re.match(_capture_regex("run", "[0-9]+", False), "123_").groupdict()
-    {'run': '123'}
+def _bidsignore_check(ignore_expression, file_name, file_root):
     """
-    return f"(?P={name})" if backref else f"(?P<{name}>{pattern})"
-
-
-def _optional_regex(regex, optional):
-    """Return an optional version of a regex if optional is True
-
-    A required regex is passed through unchanged:
-
-    >>> pattern = _optional_regex("xyz", False)
-    >>> pattern
-    'xyz'
-    >>> re.match(pattern, "xyz").groups()
-    ()
-    >>> re.match(pattern, "") is None
-    True
-
-    An optional regex uses a non-capturing group, to avoid interfering
-    with existing groups
-
-    >>> pattern = _optional_regex("x(?P<name>[a-z])z", True)
-    >>> pattern
-    '(?:x(?P<name>[a-z])z)?'
-    >>> re.match(pattern, "xyz").groups()
-    ('y',)
-    >>> re.match(pattern, "xyz").groupdict()
-    {'name': 'y'}
-    >>> re.match(pattern, "").groups()
-    (None,)
-    >>> re.match(pattern, "").groupdict()
-    {'name': None}
-    """
-    return f"(?:{regex})?" if optional else regex
+    Check whether a file is set to be ignored as per `.bidsignore`
 
+    Parameters
+    ----------
+    ignore_expression : str
+        A string following git-wildcard conventions (including `**/`).
+    file_name : str
+        A string which represents a file name.
+    file_root : str
+        The directory containing the file specifiled in `file_name`.
 
-@lru_cache()
-def _format_entity(entity, name, pattern, level, directory=False):
-    if directory and entity not in DIR_ENTITIES:
-        return ""
+    Returns
+    -------
+    bool:
+        Whether the file should be ignored.
 
-    label = _capture_regex(entity, pattern, not directory and entity in DIR_ENTITIES)
-    post = "/" if directory else "_"
+    Notes
+    -----
+    * We cannot use `glob` since that would pre-empt working with theoretical or non-local
+        paths, therefore we use `fnmatch`.
+    * `fnmatch` does not support `**/` matching as that is an optional convention from e.g.
+        globstar and git, and not part of the standard Unix shell. As we formalize `.bidsignore`
+        we may choose drop it, since we already treat simple file names as to-be-ignored in
+        all directories, and with BIDS having only up to 4 hierarchical levels, the utility of
+        other usage is limited and expansion to `..*/*..` would only mean at maximum a duplication
+        of entries.
+    """
 
-    return _optional_regex(f"{name}-{label}{post}", level != "required")
+    if ignore_expression.startswith("**/"):
+        ignore_expression = ignore_expression.lstrip("**/")
+    elif any(i in ignore_expression for i in ["/", "\\"]):
+        file_name = os.path.join(file_root, file_name)
 
+    if fnmatch.fnmatch(file_name, ignore_expression):
+        return True
+    else:
+        return False
```

Comment:
```suggestion
    return fnmatch.fnmatch(file_name, ignore_expression)
```

Reply:
done in [9c08a88](https://github.com/bids-standard/bids-specification/pull/1398/commits/9c08a88ddce61a3950dd242d0ce98596fe097046)